python-cfd/CG.py

```diff
@@ -0,0 +1,66 @@
+import sys
+import math
+import numpy as np
+import scipy as sp
+from scipy import linalg
+
+A = np.array([[-2,1,1,0],[1,-2,0,1],[1,0,-2,1],[0,1,1,-2]])
+b = np.array([1, -1, 1, -1])
+
+# Provide some information
+print("# matrix:")
+for r in A:
+    print("   ", r)
+print("  eigenvalues:", ", ".join(["{:.6g}".format(e) for e in np.linalg.eigvals(A)]))
+print("  rank:", np.linalg.matrix_rank(A))
+
+kerA = sp.linalg.null_space(A)
+print("  nullspace:")
+for r in kerA:
+    print("   ", r)
+
+# Initialize containers and values
+print("# intialization:")
+x = np.array([0., 0., 0., 0.])
+print("  initial guess:", x)
+
+residuals = [b - A @ x]
```

Comment:
we should check here is residual are small enough and exit if so

Reply:
Yes, this is a Python demonstrator but in the real C++ implementation this special case should be handled